com.unity.shadergraph/Editor/ShaderFoundry/Linker/Properties/PropertyDeclarations.cs

```diff
@@ -0,0 +1,382 @@
+                AssignmentFunctionName = "UnityBuildTexture3DStruct",
+            };
+            return HandleUnityTextureShared(context, buildingData, result);
+        }
+
+        internal class TextureBuildingData
+        {
+            internal ShaderType TextureUniformType;
+            internal string MaterialPropertyType;
+            internal string MaterialPropertyDefaultValue = "\"\" {}";
+            internal string AssignmentFunctionName;
+            internal bool HasScaleOffset = false;
+            internal bool HasTexelSize = false;
+        }
+
+        static bool HandleUnityTextureShared(FieldPropertyContext context, TextureBuildingData buildingData, FieldPropertyData result)
+        {
+            var container = context.Container;
+            var uniformName = context.UniformName;
+            result.UniformReadingData = new UniformReadingData { Rhs = $"{buildingData.AssignmentFunctionName}({uniformName})" };
+            result.UniformDeclarations.Add(new UniformDeclarationData { dataSource = context.DataSource, Name = $"{uniformName}", Type = buildingData.TextureUniformType });
+            result.UniformDeclarations.Add(new UniformDeclarationData { dataSource = context.DataSource, Name = $"sampler{uniformName}", Type = container._SamplerState });
+            if (buildingData.HasTexelSize)
+                result.UniformDeclarations.Add(new UniformDeclarationData { dataSource = context.DataSource, Name = $"{uniformName}_TexelSize", Type = container._float4 });
+            if (buildingData.HasScaleOffset)
+                result.UniformDeclarations.Add(new UniformDeclarationData { dataSource = context.DataSource, Name = $"{uniformName}_ST", Type = container._float4 });
+            MaterialPropertyDeclarationData.BuildSimple(context, buildingData.MaterialPropertyType, buildingData.MaterialPropertyDefaultValue, result);
+            return true;
+        }
+
+        static bool HandleVirtualTexture(FieldPropertyContext context, FieldPropertyData result)
+        {
+            var attributes = context.Attributes;
+            var virtualTextureAttribute = VirtualTextureAttribute.FindFirst(attributes);
+            if (virtualTextureAttribute == null)
+                return false;
+
+            var container = context.Container;
+            var uniformName = context.UniformName;
+            var displayName = context.DisplayName;
+
+            // Process all layers into a list. If a layer is not specified then it is left as null.
+            var layerCount = virtualTextureAttribute.LayerCount;
+            var layerInfo = new List<VirtualTextureLayerAttribute>(layerCount);
+            for (var i = 0; i < layerCount; ++i)
+                layerInfo.Add(null);
+            foreach (var attribute in attributes)
+            {
+                var layerAttribute = VirtualTextureLayerAttribute.TryParse(attribute);
+                if (layerAttribute != null && layerAttribute.Index < layerCount)
```

Comment:
Should we throw an error if we exceed layerCount?

Reply:
That would be good. I haven't done a serious error handling pass, partially as we were also trying to decide how to handle errors (throw exceptions?).

Reply:
I would advise against throwing exceptions where possible in Foundry.
Anything that throws an exception says "OK, I can't continue at all" and stops the whole process. If we can continue and potentially gather more errors if they are there, it will be better for the user experience.